e2e/opentelemetry/gateway.config.ts

```diff
@@ -26,9 +33,142 @@ const useOnFetchTracer = (): GatewayPlugin => {
+
+  onStart(span: Span): void {
+    const spanContext = span.spanContext();
+    const traceId = spanContext.traceId;
+    const spanId = spanContext.spanId;
+
+    // Initialize trace data structures if needed
+    if (!this.activeSpans.has(traceId)) {
+      this.activeSpans.set(traceId, new Map());
+    }
+    if (!this.subgraphNames.has(traceId)) {
+      this.subgraphNames.set(traceId, new Set());
+    }
+
+    this.activeSpans.get(traceId)!.set(spanId, span);
+
+    // If this is a root span (no parent), mark it as the root span for this trace
+    if (!span.parentSpanId) {
+      this.rootSpanIds.set(traceId, spanId);
+    }
+
+    // Check if this is a subgraph execution span
+    if (span.name && span.name.startsWith('subgraph.execute')) {
+      const subgraphName = span.attributes['gateway.upstream.subgraph.name'];
+      if (subgraphName && typeof subgraphName === 'string') {
+        this.subgraphNames.get(traceId)!.add(subgraphName);
+      }
+    }
+  }
+
+  onEnd(span: Span): void {
+    const spanContext = span.spanContext();
+    const traceId = spanContext.traceId;
+    const spanId = spanContext.spanId;
+
+    // Skip if we don't have this trace
+    if (!this.activeSpans.has(traceId)) {
+      return;
+    }
+
+    const spansForTrace = this.activeSpans.get(traceId)!;
+    const rootSpanId = this.rootSpanIds.get(traceId);
+    const subgraphNamesForTrace = this.subgraphNames.get(traceId);
+
+    // Check if this is the GraphQL execute span we're interested in
+    // TODO: can we have this fully type safe?
+    if (span.name === 'graphql.execute') {
+      const operationType = span.attributes['graphql.operation.type'];
+      const operationName = span.attributes['graphql.operation.name'];
+      const errorCount = span.attributes['graphql.error.count'];
+
+      if (rootSpanId) {
+        const rootSpan = spansForTrace.get(rootSpanId);
+        if (rootSpan && !rootSpan.ended) {
+          // Update the name of the root span
+          if (operationType && operationName) {
+            rootSpan.updateName(`${operationType} ${operationName}`);
+
+            // Copy attributes to root span
+            if (operationType)
+              rootSpan.setAttribute('graphql.operation.type', operationType);
+            if (operationName)
+              rootSpan.setAttribute('graphql.operation.name', operationName);
+            if (errorCount !== undefined)
+              rootSpan.setAttribute('graphql.error.count', errorCount);
+
+            // Add the subgraph names as a comma-separated list
+            if (subgraphNamesForTrace && subgraphNamesForTrace.size > 0) {
+              rootSpan.setAttribute(
+                'subgraph.names',
+                Array.from(subgraphNamesForTrace).join(','),
+              );
+            }
+          }
+        }
+      }
+    }
+
+    // For any subgraph span that's ending, make sure we capture its name
+    if (span.name && span.name.startsWith('subgraph.execute')) {
```

Comment:
super bad impl, we should improve and see if we can depend on the values in the gw OTEL plugin, and avoid using raw strings. also make sure we don't break it by adding tests 😆